docs: Add tutorial on encrypted events and examples for Sapphire (#612)
oasisprotocol/sapphire-paratime: docs/examples/encrypted-events.mdx

  ethers.getBytes(ciphertextFromEvent),
  aadBytes,
);

console.log('Decrypted message:', ethers.toUtf8String(plaintext));
```

This self-contained example is key to understanding the end-to-end flow without needing to run the full demo repository.

---

## Pattern 1: Passing a Symmetric Key

This is the simplest approach. The client generates a 32-byte symmetric key and passes it as an argument in the transaction to the smart contract.

### Contract Implementation

The function takes the key and the message, encrypts, and emits the event.

```solidity
import {Sapphire} from "@oasisprotocol/sapphire-contracts/contracts/Sapphire.sol";

contract EncryptedEvents {
    event Encrypted(address indexed sender, bytes32 nonce, bytes ciphertext);

    function emitEncrypted(bytes32 key, bytes calldata message) external {
        bytes32 nonce = bytes32(Sapphire.randomBytes(32, bytes("my-dapp-nonce")));
        bytes memory cipher = Sapphire.encrypt(key, nonce, message, bytes("") /* aad */);
        emit Encrypted(msg.sender, nonce, cipher);
    }
}
```

The off-chain client is then responsible for securely managing this key and using it for decryption as shown in the section above.

---

> ## Pattern 2: Deriving a Key with ECDH

Reviewer: I think something like `Deriving Symmetric Key`would be better here.

Adding a Pattern 3, which shows only onchain key generation with a guarded public function for ROFL use cases, would be helpful

### Option A: Sender-Bound AAD

Bind the encryption to the `msg.sender`. This is useful for user-specific data.

- **On-Chain AAD:** `bytes memory aad = abi.encodePacked(msg.sender);`
- **Off-Chain AAD:** Get the `sender` from the event log and convert it to bytes.

  ```typescript
  // senderAddress is from the event log
  const aadBytes = ethers.getBytes(senderAddress);
  ```

### Option B: Context-Bound AAD

Bind the encryption to the contract and chain. This is relayer-agnostic.

- **On-Chain AAD:** `bytes memory aad = abi.encodePacked(block.chainid, address(this));`
- **Off-Chain AAD:** Reconstruct the same packed bytes.

  ```typescript
  const { chainId } = await provider.getNetwork();
  const aadBytes = ethers.getBytes(
    ethers.solidityPacked(['uint256', 'address'], [chainId, contractAddress]),
  );
  ```

---

## Full Example and Demo

To see these patterns in action, you can run the [Encrypted Events Demo repository](https://github.com/oasisprotocol/encrypted-events-demo). It includes Hardhat tasks to deploy, emit, and decrypt events using both methods.

### Prerequisites

> 

Reviewer: The example should just the example admonitions
```
:::example Encrypted Events
<text>
:::
``` 
The things below should only be in the example repo, they are not relevant for the docs here.

1. **Get a Nonce:** Generate a fresh, random nonce for each encryption. Using a domain separator is a good practice.

   ```solidity
   // A 32-byte nonce for Deoxys-II, which uses the first 15 bytes
   bytes32 nonce = bytes32(Sapphire.randomBytes(32, bytes("my-dapp-nonce")));
   ```

2. **Encrypt the Data:** Use `Sapphire.encrypt` with a key, the nonce, your **text** (plaintext), and optional authenticated data (**AAD**, named `ad` in the API).

   ```solidity
   // Keep names identical to the library
   bytes memory ad = "additional data";
   bytes memory encrypted = Sapphire.encrypt(key, nonce, text, ad);
   ```

3. **Emit the Event:**

   ```solidity
   emit Encrypted(msg.sender, nonce, encrypted);
   ```

### Off-Chain Decryption

To read the secret text, an off-chain client with the correct key listens for `Encrypted` events and decrypts the `ciphertext`.

```typescript
import { AEAD, NonceSize } from '@oasisprotocol/deoxysii';
import { ethers } from 'ethers';

// You need:
// - key: A Uint8Array(32) symmetric key
// - nonceFromEvent: The 'nonce' field from the event log
// - ciphertextFromEvent: The 'ciphertext' field from the event log

> // - adBytes: AAD bytes if used on-chain (a.k.a. "authenticated data")

Reviewer 2: Why not keep it consistent with above?

```suggestion
// - ad: AAD bytes if used on-chain (a.k.a. "authenticated data")
```

const aead = new AEAD(key);

const plaintext = aead.decrypt(
  // IMPORTANT: Deoxys-II uses a 15-byte nonce.
  // We slice the first 15 bytes from the 32-byte value stored on-chain.
  ethers.getBytes(nonceFromEvent).slice(0, NonceSize),
  ethers.getBytes(ciphertextFromEvent),

>   adBytes,

Reviewer 2: ```suggestion
  ad,
```

);

console.log('Decrypted message:', ethers.toUtf8String(plaintext));
```

> ---

Reviewer 2: ```suggestion
```

## Pattern 1: Passing a Symmetric Key

> This is the simplest approach. The client generates a 32‑byte symmetric key and passes it as an argument in the (Sapphire‑encrypted) transaction to the smart contract.

Reviewer 2: ```suggestion
This is the simplest approach. The client generates a 32‑byte symmetric key and passes it to the smart contract as an argument inside the encrypted transaction.
```

### TypeScript: generate a `bytes32` key

> ```ts

Reviewer 2: We can be fancy here and use TabWidget here.

import { ethers } from 'ethers';
import { randomBytes } from 'crypto';

// Option A (ethers v6)
const keyHex = ethers.hexlify(ethers.randomBytes(32)) as `0x${string}`;

// Option B (Node.js crypto)

> const keyHex2 = ('0x' + randomBytes(32).toString('hex')) as `0x${string}`;

Reviewer 2: Is `0x` redundant here?

Author: ethers v6 expects hex strings to be 0x prefixed. replaced with `ethers.hexlify`

// Pass keyHex/keyHex2 directly to a Solidity function taking `bytes32 key`.
```

### Contract Implementation

> The function takes the key and the **text**, encrypts, and emits the event.

Reviewer 2: ```suggestion
The function takes the key and the text, encrypts, and emits the event.
```

```solidity

> import {Sapphire} from "@oasisprotocol/sapphire-contracts/contracts/Sapphire.sol";

Reviewer 2: ```suggestion
import { Sapphire } from "@oasisprotocol/sapphire-contracts/contracts/Sapphire.sol";
```

        bytes memory encrypted = Sapphire.encrypt(key, nonce, text, ad);
        emit Encrypted(msg.sender, nonce, encrypted);
    }
}
```

The off-chain client is responsible for managing this key and using it for decryption as shown above.

---

## Pattern 2: Deriving a Symmetric Key (ECDH)

In this more advanced pattern, you don't send a key with every transaction. Instead, the contract holds a long‑lived Curve25519 secret key (in confidential state), and a shared secret is **derived on‑chain** from the caller’s public key using X25519 (ECDH).

### Contract Implementation (ECDH)

1. **Generate a Keypair:** In the constructor, generate a Curve25519 keypair and store the secret key in confidential state. The public key can be exposed for clients.

   ```solidity
   contract EncryptedEventsECDH {
       Sapphire.Curve25519SecretKey private _contractSk;
       Sapphire.Curve25519PublicKey public contractPk;

       constructor() {
           (Sapphire.Curve25519PublicKey pk, Sapphire.Curve25519SecretKey sk) =
               Sapphire.generateCurve25519KeyPair(bytes(""));
           contractPk = pk;
           _contractSk = sk;
       }
       // ...
   }
   ```

> 2. **Derive and Encrypt:** The emit function now takes the caller's public key, derives the shared key, encrypts the **text**, and emits the event.

Reviewer 2: ```suggestion
2. **Derive and Encrypt:** The emit function now takes the caller's public key, derives the shared key, encrypts the text, and emits the event.
```

   ```solidity
   function emitEncryptedECDH(
       Sapphire.Curve25519PublicKey callerPublicKey,
       bytes calldata text
   ) external {
       // Derive the shared symmetric key
       bytes32 key = Sapphire.deriveSymmetricKey(callerPublicKey, _contractSk);

       // Encrypt and emit as before
       bytes32 nonce = bytes32(Sapphire.randomBytes(32, bytes("my-dapp-nonce")));
       bytes memory ad = bytes(""); // optional AAD
       bytes memory encrypted = Sapphire.encrypt(key, nonce, text, ad);
       emit Encrypted(msg.sender, nonce, encrypted);
   }
   ```

### Off-Chain Key Derivation

> The client derives the same key from its **secret** and the contract’s **public** key.

Reviewer 2: ```suggestion
The client derives the shared key from its **secret** and the contract’s **public** key.
```

```typescript
import { mraeDeoxysii } from '@oasisprotocol/client-rt';
import { AEAD } from '@oasisprotocol/deoxysii';
import { ethers } from 'ethers';

// contractPkHex: 0x-prefixed Curve25519 public key (bytes32) fetched on-chain
// callerSkBytes: Uint8Array(32) caller's Curve25519 secret key
const sharedKey = mraeDeoxysii.deriveSymmetricKey(
  ethers.getBytes(contractPkHex),
  callerSkBytes,
);

const aead = new AEAD(sharedKey);

> // ... decrypt as shown earlier (using adBytes if AAD was used) ...

Reviewer 2: I needed a while to find the place. Mentioning `aead.decrypt()` would help.

```suggestion
// ... decrypt with aead.decrypt() as shown earlier.
```

```

> ---

Reviewer 2: ```suggestion
```
